File: backend/crud/agents.py
"""
CRUD operations for Agent entities.

This module contains pure database operations only.
Business logic (config loading, prompt building) is in services/agent_factory.py.
"""

import logging
from typing import List, Optional

import models
from database import retry_on_db_lock, serialized_write
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select

logger = logging.getLogger("CRUD")


@retry_on_db_lock(max_retries=5, initial_delay=0.1, backoff_factor=2)
async def create_agent(
    db: AsyncSession,
    name: str,
    system_prompt: str,
    profile_pic: Optional[str] = None,
    in_a_nutshell: Optional[str] = None,
    characteristics: Optional[str] = None,
    recent_events: Optional[str] = None,
    group: Optional[str] = None,
    config_file: Optional[str] = None,
    interrupt_every_turn: bool = False,
    priority: int = 0,
    transparent: bool = False,
    world_name: Optional[str] = None,
) -> models.Agent:
    """
    Create an agent in the database.

    This is a pure CRUD operation. For creating agents from config files,
    use AgentFactory.create_from_config() instead.

    Args:
        db: Database session
        name: Agent name
        system_prompt: Pre-built system prompt
        profile_pic: Profile picture URL or path
        in_a_nutshell: Brief agent summary
        characteristics: Agent personality traits
        recent_events: Recent events text
        group: Optional group name
        config_file: Optional config file path
        interrupt_every_turn: Whether agent interrupts every turn
        priority: Agent priority (higher = responds first)
        transparent: Whether agent's responses don't trigger others
        world_name: World name for world-specific characters (NULL for system agents)

    Returns:
        Created Agent model
    """
    db_agent = models.Agent(
        name=name,
        world_name=world_name,
        group=group,
        config_file=config_file,
        profile_pic=profile_pic,
        in_a_nutshell=in_a_nutshell,
        characteristics=characteristics,
        recent_events=recent_events,
        system_prompt=system_prompt,
        interrupt_every_turn=bool(interrupt_every_turn),
        priority=priority,
        transparent=bool(transparent),
    )
    db.add(db_agent)
    async with serialized_write():
        await db.commit()
    await db.refresh(db_agent)
    return db_agent


async def get_all_agents(db: AsyncSession) -> List[models.Agent]:
    """Get all agents globally."""
    result = await db.execute(select(models.Agent))
    return result.scalars().all()


async def get_agent(db: AsyncSession, agent_id: int) -> Optional[models.Agent]:
    """Get a specific agent by ID."""
    result = await db.execute(select(models.Agent).where(models.Agent.id == agent_id))
    return result.scalar_one_or_none()


async def get_agent_by_name(db: AsyncSession, name: str, world_name: Optional[str] = None) -> Optional[models.Agent]:
    """
    Get a specific agent by name, handling whitespace/underscore variations.

    Args:
        db: Database session
        name: Agent name to search for
        world_name: Optional world name to filter by (required when multiple
                    agents with same name exist across different worlds)

    Tries multiple name formats:
    1. Original name as-is
    2. Spaces replaced with underscores
    3. Underscores replaced with spaces
    """

    def _build_query(agent_name: str):
        """Build query with optional world_name filter."""
        query = select(models.Agent).where(models.Agent.name == agent_name)
        if world_name is not None:
            query = query.where(models.Agent.world_name == world_name)
        return query

    # Try original name first
    result = await db.execute(_build_query(name))
    agent = result.scalar_one_or_none()
    if agent:
        return agent

    # Try with spaces -> underscores
    name_with_underscores = name.replace(" ", "_")
    if name_with_underscores != name:
        result = await db.execute(_build_query(name_with_underscores))
        agent = result.scalar_one_or_none()
        if agent:
            return agent

    # Try with underscores -> spaces
    name_with_spaces = name.replace("_", " ")
    if name_with_spaces != name:
        result = await db.execute(_build_query(name_with_spaces))
        agent = result.scalar_one_or_none()
        if agent:
            return agent

    return None


@retry_on_db_lock(max_retries=5, initial_delay=0.1, backoff_factor=2)
async def delete_agent(db: AsyncSession, agent_id: int) -> bool:
    """Delete an agent permanently."""
    result = await db.execute(select(models.Agent).where(models.Agent.id == agent_id))
    agent = result.scalar_one_or_none()
    if agent:
        await db.delete(agent)
        async with serialized_write():
            await db.commit()
        return True
    return False


@retry_on_db_lock(max_retries=5, initial_delay=0.1, backoff_factor=2)
async def delete_agents_by_world(db: AsyncSession, world_name: str) -> int:
    """
    Delete all agents associated with a specific world.

    Args:
        db: Database session
        world_name: The world name to delete agents for

    Returns:
        Number of agents deleted
    """
    result = await db.execute(select(models.Agent).where(models.Agent.world_name == world_name))
    agents = result.scalars().all()

    count = len(agents)
    for agent in agents:
        await db.delete(agent)

    if count > 0:
        async with serialized_write():
            await db.commit()
        logger.info(f"Deleted {count} agents for world '{world_name}'")

    return count


async def get_agents_by_world(db: AsyncSession, world_name: str) -> List[models.Agent]:
    """Get all agents for a specific world."""
    result = await db.execute(select(models.Agent).where(models.Agent.world_name == world_name))
    return list(result.scalars().all())


@retry_on_db_lock(max_retries=5, initial_delay=0.1, backoff_factor=2)
async def sync_agents_with_filesystem(db: AsyncSession, world_name: str) -> int:
    """
    Sync agents with filesystem - delete agents whose config directories no longer exist.

    This handles the case where agent folders are removed from the filesystem
    but database entries remain (stale agents).

    Args:
        db: Database session
        world_name: The world name to sync agents for

    Returns:
        Number of stale agents deleted
    """
    from pathlib import Path

    agents = await get_agents_by_world(db, world_name)
    deleted_count = 0

    for agent in agents:
        if agent.config_file:
            config_path = Path(agent.config_file)
            # Check if config directory exists (for folder-based configs)
            # or if config file exists (for single-file configs)
            if not config_path.exists() and not config_path.with_suffix(".md").exists():
                logger.info(f"Deleting stale agent '{agent.name}' - config not found at '{agent.config_file}'")
                await db.delete(agent)
                deleted_count += 1

    if deleted_count > 0:
        async with serialized_write():
            await db.commit()
        logger.info(f"Cleaned up {deleted_count} stale agents for world '{world_name}'")

    return deleted_count


@retry_on_db_lock(max_retries=5, initial_delay=0.1, backoff_factor=2)
async def update_agent(
    db: AsyncSession,
    agent_id: int,
    system_prompt: Optional[str] = None,
    profile_pic: Optional[str] = None,
    in_a_nutshell: Optional[str] = None,
    characteristics: Optional[str] = None,
    recent_events: Optional[str] = None,
    interrupt_every_turn: Optional[bool] = None,
    priority: Optional[int] = None,
    transparent: Optional[bool] = None,
) -> Optional[models.Agent]:
    """
    Update an agent's fields in the database.

    This is a pure CRUD operation. For reloading from config files,
    use AgentFactory.reload_from_config() instead.

    Only updates fields that are explicitly provided (not None).

    Args:
        db: Database session
        agent_id: Agent ID
        system_prompt: New system prompt
        profile_pic: New profile picture
        in_a_nutshell: New nutshell summary
        characteristics: New characteristics
        recent_events: New recent events
        interrupt_every_turn: Whether agent interrupts every turn
        priority: Agent priority
        transparent: Whether agent is transparent

    Returns:
        Updated Agent model or None if not found
    """
    result = await db.execute(select(models.Agent).where(models.Agent.id == agent_id))
    agent = result.scalar_one_or_none()

    if not agent:
        return None

    # Update fields if provided
    if system_prompt is not None:
        agent.system_prompt = system_prompt
    if profile_pic is not None:
        # Check if it's a base64 data URL - save to filesystem
        if profile_pic.startswith("data:image/"):
            from services.agent_config_service import AgentConfigService

            AgentConfigService.save_base64_profile_pic(agent.name, profile_pic)
            agent.profile_pic = None  # Clear DB field - images served from filesystem
        else:
            agent.profile_pic = profile_pic
    if in_a_nutshell is not None:
        agent.in_a_nutshell = in_a_nutshell
    if characteristics is not None:
        agent.characteristics = characteristics
    if recent_events is not None:
        agent.recent_events = recent_events
    if interrupt_every_turn is not None:
        agent.interrupt_every_turn = bool(interrupt_every_turn)
    if priority is not None:
        agent.priority = priority
    if transparent is not None:
        agent.transparent = bool(transparent)

    async with serialized_write():
        await db.commit()
    await db.refresh(agent)

    # Invalidate agent caches
    from infrastructure.cache import agent_config_key, agent_object_key, get_cache

    cache = get_cache()
    cache.invalidate(agent_config_key(agent_id))
    cache.invalidate(agent_object_key(agent_id))

    return agent

File: backend/src/crud/agents.ts
/** CRUD operations for Agent entities. */

import { and, eq, inArray } from 'drizzle-orm'
import { existsSync } from 'node:fs'
import { basename, dirname, isAbsolute, join } from 'node:path'
import { createProjectPaths } from '@/config/paths'
import type { Db } from '@/db'
import { agents, roomAgents, type Agent } from '@/db/schema'
import { getCache, roomAgentsKey } from '@/infrastructure/cache'
import { getLogger } from '@/infrastructure/logging/logger'
import { invalidateAgentCache } from './cache-invalidation'

const logger = getLogger('CRUD')

export interface AgentCreate {
  name: string
  systemPrompt: string
  profilePic?: string | null
  inANutshell?: string | null
  characteristics?: string | null
  recentEvents?: string | null
  group?: string | null
  configFile?: string | null
  interruptEveryTurn?: boolean
  priority?: number
  transparent?: boolean
  /** NULL for system agents; a world name scopes the agent to one world's cast. */
  worldName?: string | null
}

// The booleans, `priority` and `created_at` carry no SQL DEFAULT, and NULL and
// 0 sort differently in the ordering `priority` feeds.
export function createAgent(db: Db, agent: AgentCreate): Agent {
  return db
    .insert(agents)
    .values({
      name: agent.name,
      worldName: agent.worldName ?? null,
      group: agent.group ?? null,
      configFile: agent.configFile ?? null,
      profilePic: agent.profilePic ?? null,
      inANutshell: agent.inANutshell ?? null,
      characteristics: agent.characteristics ?? null,
      recentEvents: agent.recentEvents ?? null,
      systemPrompt: agent.systemPrompt,
      interruptEveryTurn: agent.interruptEveryTurn ?? false,
      priority: agent.priority ?? 0,
      transparent: agent.transparent ?? false,
      createdAt: new Date(),
    })
    .returning()
    .get()
}

// Unlike `LocationUpdate`, `null` and `undefined` both mean "skip" here.
export interface AgentUpdate {
  systemPrompt?: string | null
  profilePic?: string | null
  inANutshell?: string | null
  characteristics?: string | null
  recentEvents?: string | null
  interruptEveryTurn?: boolean | null
  priority?: number | null
  transparent?: boolean | null
}

export function updateAgent(db: Db, agentId: number, update: AgentUpdate): Agent | null {
  const existing = db.select().from(agents).where(eq(agents.id, agentId)).get()
  if (!existing) return null

  const patch: Partial<typeof agents.$inferInsert> = {}

  if (update.systemPrompt != null) patch.systemPrompt = update.systemPrompt
  if (update.profilePic != null) {
    if (update.profilePic.startsWith('data:image/')) {
      // TODO(phase-3): write the image to `agents/<name>/profile.<ext>` before
      // nulling the column — `services/agent-config.ts` has no writer yet, so
      // the upload is accepted and the bytes go nowhere. Hence the error log.
      logger.error(
        `Discarded a base64 profile picture for agent '${existing.name}': the filesystem ` +
          'writer (AgentConfigService.save_base64_profile_pic) is not ported yet.',
      )
      patch.profilePic = null
    } else {
      patch.profilePic = update.profilePic
    }
  }
  if (update.inANutshell != null) patch.inANutshell = update.inANutshell
  if (update.characteristics != null) patch.characteristics = update.characteristics
  if (update.recentEvents != null) patch.recentEvents = update.recentEvents
  if (update.interruptEveryTurn != null) patch.interruptEveryTurn = update.interruptEveryTurn
  if (update.priority != null) patch.priority = update.priority
  if (update.transparent != null) patch.transparent = update.transparent

  // Drizzle rejects an empty SET clause.
  const updated =
    Object.keys(patch).length === 0
      ? existing
      : db.update(agents).set(patch).where(eq(agents.id, agentId)).returning().get()

  // The SDK layer reads the id-keyed prompt cache at the start of every turn, so
  // a system-prompt edit not followed by this stays invisible until it expires.
  invalidateAgentCache(agentId)

  return updated
}

export function getAgent(db: Db, agentId: number): Agent | null {
  return db.select().from(agents).where(eq(agents.id, agentId)).get() ?? null
}

// Unscoped: system agents (NULL `world_name`) mixed with every world's cast.
export function getAllAgents(db: Db): Agent[] {
  return db.select().from(agents).all()
}

// `room_agents` and `room_agent_sessions` cascade; `messages.agent_id` is
// `ON DELETE SET NULL`. Both cache sweeps are required: SQLite reuses rowids, so
// a cached `agent_config:{id}` can reach a *different* agent that inherits the
// id, and without the membership sweep the orchestrator keeps giving it turns.
export function deleteAgent(db: Db, agentId: number): boolean {
  const memberships = db
    .select({ roomId: roomAgents.roomId })
    .from(roomAgents)
    .where(eq(roomAgents.agentId, agentId))
    .all()

  const deleted = db.delete(agents).where(eq(agents.id, agentId)).returning({ id: agents.id }).get()
  if (!deleted) return false

  invalidateAgentCache(agentId)
  for (const { roomId } of memberships) getCache().invalidate(roomAgentsKey(roomId))

  return true
}

/**
 * Get an agent by name, tolerating whitespace/underscore variations. The three
 * spellings are probed in a fixed order, and the order is load-bearing: a world
 * can hold both "Foo Bar" and "Foo_Bar". `worldName` filters only when
 * supplied, and a NULL one is reachable *only* by omitting the argument.
 */
export function getAgentByName(db: Db, name: string, worldName?: string | null): Agent | null {
  const lookup = (candidate: string): Agent | null => {
    const nameMatch = eq(agents.name, candidate)
    const where =
      worldName === undefined || worldName === null
        ? nameMatch
        : and(nameMatch, eq(agents.worldName, worldName))
    return db.select().from(agents).where(where).get() ?? null
  }

  // Variants equal to the original are dropped rather than re-queried.
  const candidates = [name, name.replaceAll(' ', '_'), name.replaceAll('_', ' ')].filter(
    (candidate, i) => i === 0 || candidate !== name,
  )

  for (const candidate of candidates) {
    const found = lookup(candidate)
    if (found) return found
  }
  return null
}

// System agents are excluded by construction: `world_name` is NULL for them.
export function getAgentsByWorld(db: Db, worldName: string): Agent[] {
  return db.select().from(agents).where(eq(agents.worldName, worldName)).all()
}

// Replaces an existing extension rather than appending; a leading dot is part
// of the name (`.hidden` → `.hidden.md`).
function withMarkdownSuffix(path: string): string {
  const dir = dirname(path)
  const base = basename(path)
  const dot = base.lastIndexOf('.')
  const stem = dot > 0 ? base.slice(0, dot) : base
  return join(dir, `${stem}.md`)
}

export interface SyncAgentsOptions {
  /** Root relative `config_file` values resolve against; defaults to the project root. */
  projectRoot?: string
}

/**
 * Delete agent rows whose config no longer exists on disk; the filesystem wins.
 * Stored `config_file` values are project-root relative and must be resolved
 * against the project root, never the cwd — against the cwd every probe misses
 * and the world's entire cast is deleted. A row with no `config_file` is left
 * alone: nothing on disk claims it.
 */
export function syncAgentsWithFilesystem(
  db: Db,
  worldName: string,
  { projectRoot }: SyncAgentsOptions = {},
): number {
  const root = projectRoot ?? createProjectPaths().projectRoot

  const stale: Agent[] = []
  for (const agent of getAgentsByWorld(db, worldName)) {
    if (!agent.configFile) continue

    const configPath = isAbsolute(agent.configFile) ? agent.configFile : join(root, agent.configFile)
    // Two shapes are legal: a folder of `.md` files, or a single `.md` file.
    if (existsSync(configPath) || existsSync(withMarkdownSuffix(configPath))) continue

    logger.info(
      `Deleting stale agent '${agent.name}' - config not found at '${agent.configFile}'`,
    )
    stale.push(agent)
  }

  if (stale.length === 0) return 0

  // One statement, so a failure leaves the cast intact rather than half-removed.
  db.delete(agents)
    .where(
      inArray(
        agents.id,
        stale.map((agent) => agent.id),
      ),
    )
    .run()

  logger.info(`Cleaned up ${stale.length} stale agents for world '${worldName}'`)
  return stale.length
}
